scipy/stats/_hypotests.py

```diff
@@ -144,6 +144,222 @@ def epps_singleton_2samp(x, y, t=(0.4, 0.8)):
+    # to compute the p-value". [1] page 26 below eq. (3.6).
+    if lmbd_hat2 <= 0:
+        return PoissonMeansTestResult(0, 1)
+    # the unbiased variance estimate [1] (3.2)
+    var = k1 / (n1 ** 2) + k2 / (n2 ** 2)
+
+    # the _observed_ pivot statistic from the input. It follows the
+    # unnumbered equation following equation (3.3) This is used later in
+    # comparison with the computed pivot statistics in an indicator function.
+    t_k1k2 = (k1 / n1 - k2 / n2 - diff) / np.sqrt(var)
+    # equation (3.5) of [1] is lengthy, so it is broken into several parts,
+    # beginning here. Note that the probability mass function of poisson is
+    # exp^(-\mu)*\mu^k/k!, so and this is called with shape \mu, here noted
+    # here as nlmbd_hat*. The strategy for evaluating the double summation in
+    # (3.5) is to create two arrays of the values of the two products inside
+    # the summation and then broadcast them together into a matrix, and then
+    # sum across the entire matrix.
+
+    # compute constants (as seen in the first and second separated products in
+    # (3.5).). (This is the shape (\mu) parameter of the poisson distribution.)
+    nlmbd_hat1 = n1 * (lmbd_hat2 + diff)
+    nlmbd_hat2 = n2 * lmbd_hat2
+
+    # determine summation bounds for tail ends of distribution rather than
+    # summing to infinity. `x1*` is for the outer sum and `x2*` is the inner
+    # sum
+    x1_lb, x1_ub = distributions.poisson.ppf([1e-10, 1 - 1e-16], nlmbd_hat1)
+    x2_lb, x2_ub = distributions.poisson.ppf([1e-10, 1 - 1e-16], nlmbd_hat2)
+
+    # construct arrays to function as the x_1 and x_2 counters on the summation
+    # in (3.5). `x1` is in columns and `x2` is in rows to allow for
+    # broadcasting.
+    x1 = np.arange(x1_lb, x1_ub + 1)
+    x2 = np.arange(x2_lb, x2_ub + 1)[:, None]
+
+    # these are the two products in equation (3.5) with `prob_x1` being the
+    # first (left side) and `prob_x2` being the second (right side). (To
+    # make as clear as possible: the 1st contains a "+ d" term, the 2nd does
+    # not.)
+    prob_x1 = distributions.poisson.pmf(x1, nlmbd_hat1)
+    prob_x2 = distributions.poisson.pmf(x2, nlmbd_hat2)
+
+    # compute constants for use in the the "pivot statistic" per the
+    # unnumbered equation following (3.3).
+    lmbd_x1 = x1 / n1
+    lmbd_x2 = x2 / n2
+    lmbds_diff = lmbd_x1 - lmbd_x2 - diff
+    var_x1x2 = lmbd_x1 / n1 + lmbd_x2 / n2
+
+    # this is the 'pivot statistic' for use in the indicator of the summation
+    # (left side of "I[.]"). Before dividing, mask zero-elements in the
+    # denominator with infinity so that they are `false` in the indicator.
+    if alternative == 'two-sided':
+        mask_out_invalid = np.abs(lmbd_x1 - lmbd_x2) > diff
+    elif alternative == 'less':
+        mask_out_invalid = lmbds_diff < 0
+    else:
+        mask_out_invalid = lmbds_diff > 0
+
+    var_x1x2[~mask_out_invalid] = np.inf
```

Comment:
@swallan I am working on getting this merged, but I need a little bit of your time to finish it. I think I only have one question - where did this come from? I don't see this in the paper, and without it, we get the symmetry property I was expecting. I see that removing it makes one of the `test_fortran_authors` tests fail, but my guess is that the test (perhaps the authors' code) is wrong about this. Also, it looks like statsmodels really has this test now ([here](https://github.com/statsmodels/statsmodels/blob/2eac8066b068a88f00a29f8ff728b04e58248375/statsmodels/stats/rates.py#L1031-L1071)), and I'm not seeing this sort of condition there.

Reply:
I'm pretty confident that this `mask_out_invalid` stuff was not intended. 

If we perform the test under the null hypothesis, the p-value of the test should be uniformly distributed on the interval [0, 1] (i.e. we should get p-value < 0.01 about 1% of the time, 0.05 about 5% of the time, etc.). We can test this with an experiment that repeatedly samples from two Poisson distributions with $\lambda_1 = \lambda_2 + d$, where $d$ is the hypothesized difference. Here, I do that and I plot the empirical CDF of the observed p-values, which we expect to be a straight line with slope 1 if the p-values follow a uniform distribution.

![image](https://user-images.githubusercontent.com/6570539/182198469-7840f9f8-6064-4896-baee-1deaa8686567.png)

For instance, with `mask_out_invalid`, 1130 of the 10000 p-values were <0.05 when the null hypothesis was true, whereas without the `mask_out_invalid` code, 476 of the 10000 p-values were <0.05 when the null hypothesis was true. So without the `mask_out_invalid` code, the test controls the Type I error rate, but with the code, the test does not. I'll go ahead and remove it.

<details>

```python3
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats

N = 10000

n1 = 1000
n2 = 1000

diff = 0.01
lam2 = 0.01
lam1 = lam2 + diff

dist1 = stats.poisson(n1*lam1)
dist2 = stats.poisson(n2*lam2)

pvals = []
for i in range(N):
    k1 = dist1.rvs()
    k2 = dist2.rvs()
    res = stats.poisson_means_test(k1, n1, k2, n2, diff=diff)
    pvals.append(res.pvalue)

pvals = np.sort(pvals)
ecdf = np.linspace(0, 1, N+1)
plt.plot(ecdf, ecdf)
plt.plot(pvals, ecdf[1:])
plt.axis('square')

```

</details>

@tupui @Micky774 thought you'd like this based on our conversation at SciPy : )

Reply:
```suggestion
    # this is the 'pivot statistic' for use in the indicator of the summation
    # (left side of "I[.]").
```

Reply:
> @tupui @Micky774 thought you'd like this based on our conversation at SciPy : )

Oh this is a fantastic, thanks for pinging me on this :). I'm excited to show my students an "in the wild" use case of this concept.

Reply:
In the initial submission, there was this same calculation using `np.divide` with a `where`:

https://github.com/swallan/scipy/commit/a3f3a14e61f192294c5bcec14d223362c2ec810b#diff-6ba7fcf5a871c6742dd082e5fc1801ef3ce2126456b052e9d0a126e70aab5f99L352-L354

with the description 

>         # exclude the same cases that will be excluded in the indicator in the
>         # following `np.multiply` call to avoid invalid divisions. In these
>         # for 'skipped' indices, the result is zero, as seen in the `out=`

So that's where it came from. As to whether it is necessary, I don't know yet. I'm looking at the paper and the implementation to be sure.

Reply:
I see. OK, well take a look to see if you can find it. If you're also not sure where it's coming from, go ahead and commit the suggestions. If you're happy with the other changes, I'll go ahead and merge!